kmk/extensions/oled.py

```diff
@@ -0,0 +1,291 @@
+            _x_anchor = 0.0
+        if y_anchor == 'top' or y_anchor == 'T':
+            _y_anchor = 0.0
+        elif y_anchor == 'middle' or y_anchor == 'M':
+            _y_anchor = 0.5
+        elif y_anchor == 'bottom' or y_anchor == 'B':
+            _y_anchor = 1.0
+        else:
+            _y_anchor = 0.0
+        if (
+            x_anchor != 'middle'
+            or x_anchor != 'right'
+            or x_anchor != 'm'
+            or x_anchor != 'r'
+        ):
+            _x = x + 1
+        else:
+            _x = x
+        return {
+            0: text,
+            1: _x,
+            2: y,
+            3: side,
+            4: layer,
+            5: OledEntryType.TXT,
+            6: _x_anchor,
+            7: _y_anchor,
+            8: direction,
+            9: line_spacing,
+            10: inverted,
+        }
+
+    @staticmethod
+    def oled_image_entry(x=0, y=0, image='', layer=None, side=None):
+        odb = displayio.OnDiskBitmap(image)
+        return {
+            0: odb,
+            1: x,
+            2: y,
+            3: side,
+            4: layer,
+            5: OledEntryType.IMG,
+        }
+
+
+class Oled(Extension):
+    def __init__(
+        self,
+        views,
+        width=128,
+        height=32,
+        split=None,
+        flip: bool = False,
+        flip_left: bool = False,
+        flip_right: bool = False,
+        device_address=0x3C,
+        brightness=0.8,
+        brightness_step=0.1,
+        dim_time=None,
+        off_time=None,
+        powersave_dim_time=10,
+        powersave_off_time=60,
+    ):
+        displayio.release_displays()
+        self._flip = flip
+        self._flip_left = flip_left
+        self._flip_right = flip_right
+        self._views = views.data
+        self._width = width
+        self._height = height
+        self._prevLayers = 0
+        self._device_address = device_address
+        self._brightness = brightness
+        self._brightness_step = brightness_step
+        self._timer_start = ticks_ms()
+        self._dark = False
+        self._go_dark = False
+        self._powersave = False
+        self._dim_time_ms = dim_time * 1000 if dim_time else None
+        self._off_time_ms = off_time * 1000 if off_time else None
+        self._powersave_dim_time_ms = (
+            powersave_dim_time * 1000 if powersave_dim_time else None
+        )
+        self._powersave_off_time_ms = (
+            powersave_off_time * 1000 if powersave_off_time else None
+        )
+        self._split = split
+
+        make_key(
+            names=('OLED_BRI',), on_press=self._oled_bri, on_release=handler_passthrough
+        )
+        make_key(
+            names=('OLED_BRD',), on_press=self._oled_brd, on_release=handler_passthrough
+        )
+
+    def render_oled(self, layer, *args, **kwargs):
+        splash = displayio.Group()
+
+        for view in self._views:
+            if self._dark is False:
+                if view[4] == layer or view[4] is None:
```

Comment:
Flatten out those guards. General rule of thumb: if your indentation has more than 4 levels, it probably needs a refactor.
```suggestion
            if self._dark:
                return
            if view[4] is not None or view[4] != layer:
                return
```

Reply:
That's something I just learnt, thanks!

Reply:
I managed to get rid of ```self._dark``` and ```self._go_dark``` and all the associated checks.
The
```
            if view[4] is not None or view[4] != layer:
                return
```
Doesn't really work as I want it to keep going through ```self._views``` even if the clause returns ```False```